internal/configmodels/configmodels.go

```diff
@@ -0,0 +1,157 @@
+// Copyright 2019, OpenCensus Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package configmodels defines the data models for entities. This file defines the
+// models for V2 configuration format. The defined entities are:
+// ConfigV2 (the top-level structure), Receivers, Exporters, Processors, PipelinesCfg.
+package configmodels
+
+/*
+Receivers, Exporters and Processors typically have common configuration settings, however
+sometimes specific implementations (e.g. StackDriver) will have extra configuration settings.
+This requires the configuration data for these entities to be polymorphic.
+
+To satisfy these requirements we declare interfaces ReceiverCfg, ExporterCfg, ProcessorCfg,
+which define the behavior. We also provide helper structs ReceiverCommon, ExporterCommon,
+ProcessorCommon, which define the common settings and un-marshaling from config files.
+
+Specific Receivers/Exporters/Processors are expected to at the minimum implement the
+corresponding interface and if they have additional settings they must also extend
+the corresponding common settings struct (the easiest approach is to embed the common struct).
+*/
+
+// ConfigV2 defines the configuration V2 for the various elements of collector or agent.
+type ConfigV2 struct {
+	Receivers  []ReceiverCfg  `mapstructure:"-"`
+	Exporters  []ExporterCfg  `mapstructure:"-"`
+	Processors []ProcessorCfg `mapstructure:"-"`
+	Pipelines  PipelinesCfg   `mapstructure:"pipelines"`
+}
+
+// ReceiversKeyName is the configuration key name for receivers section.
+const ReceiversKeyName = "receivers"
+
+// ExportersKeyName is the configuration key name for exporters section.
+const ExportersKeyName = "exporters"
+
+// ProcessorsKeyName is the configuration key name for processors section.
+const ProcessorsKeyName = "processors"
+
+// PipelinesCfg defines the pipelines section of configuration.
+type PipelinesCfg struct {
+	Traces  []*PipelineCfg `mapstructure:"traces"`
+	Metrics []*PipelineCfg `mapstructure:"metrics"`
+}
+
+// PipelineCfg defines a single pipeline.
+type PipelineCfg struct {
+	Name       string              `mapstructure:"name"`
+	Receivers  []PipelineReceiver  `mapstructure:"receivers"`
+	Processors []PipelineProcessor `mapstructure:"ordered-processors"`
+	Exporters  []PipelineExporter  `mapstructure:"exporters"`
+}
+
+// PipelineReceiver defines a receiver of a pipeline.
+type PipelineReceiver struct {
```

Comment:
> ```go
> // ReceiversCfg defines the receivers section of configuration.
> type ReceiversCfg struct {
> 	TracesReceiversCfg []*ReceiverCfg `mapstructure:"traces"`
> 	MetricsReceiversCfg []*ReceiverCfg `mapstructure:"metrics"`
> }
> ```
> 

This implies that `receivers` top-level key will have 2 sub-keys `traces` and `metrics`, i.e. the entire config will look like this:

```yaml
receivers:
  traces:
    - type: zipkin
      endpoint: "127.0.0.1:12345"

  metrics:
    - type: prometheus
      endpoint: "127.0.0.1:5678"

processors:
  - type: batch

exporters:
  - type: jaeger
    enabled: true

pipelines:
  traces:
    - name: default
      receivers:
        - name: zipkin
      ordered-processors:
        - name: batch
      exporters:
        - name: jaeger
  metrics:
```

This is different from what we have in the spec. Is this what you are thinking about?

Reply:
> Maybe suffix with Cfg for consistency, also to ensure that the name reflects they are configs and not parts of the Pipeline.
> 
> Alternative suggestion is to drop Cfg from everywhere and rely on the package name to suggest that they are config models.

Actually, this is an interesting suggestion, let me try in the code to see how it looks.

Reply:
I think GitHub messed all the comments hierarchy. Related to having traces/metrics as sub-levels for the exporter, receiver, processor, I don't remember if we made that decision because we had arguments, or we missed that, or why we did not have it :)

@pjanotti @sjkaris @rghetia @songy23 do you remember why?

Reply:
>I think GitHub messed all the comments hierarchy.

Looks like a Github bug. Use the "Files" view, it seems to be correct: https://github.com/census-instrumentation/opencensus-service/pull/549/files

Reply:
Let me know the result of 

> Actually, this is an interesting suggestion, let me try in the code to see how it looks.

Reply:
Yes, I think you are right, Cfg suffix is unneeded. I think it looks readable without it:

```go
type ConfigV2 struct {
	Receivers  []Receiver  `mapstructure:"-"`
	Exporters  []Exporter  `mapstructure:"-"`
	Processors []Processor `mapstructure:"-"`
	Pipelines  Pipelines   `mapstructure:"pipelines"`
}

// Pipelines defines the pipelines section of configuration.
type Pipelines struct {
	Traces  []*Pipeline `mapstructure:"traces"`
	Metrics []*Pipeline `mapstructure:"metrics"`
}

// Pipeline defines a single pipeline.
type Pipeline struct {
	Name       string              `mapstructure:"name"`
	Receivers  []PipelineReceiver  `mapstructure:"receivers"`
	Processors []PipelineProcessor `mapstructure:"ordered-processors"`
	Exporters  []PipelineExporter  `mapstructure:"exporters"`
}

// PipelineReceiver defines a receiver of a pipeline.
type PipelineReceiver struct {
	Name string `mapstructure:"name"`
}

// PipelineProcessor defines a processor of a pipeline.
type PipelineProcessor struct {
	Name string `mapstructure:"name"`
}

// PipelineExporter defines an exporter of a pipeline.
type PipelineExporter struct {
	Name string `mapstructure:"name"`
}
```

Thoughts?

Reply:
> Related to having traces/metrics as sub-levels for the exporter, receiver, processor, I don't remember if we made that decision because we had arguments, or we missed that, or why we did not have it :)
> 
> @pjanotti @sjkaris @rghetia @songy23 do you remember why?

I think your original proposal (without sublevels for the exporter, receiver, processor) looks nicer. For receivers which support both metrics and traces it allows defining the receiver using one config section. 

One benefit when using sublevels is that you can enable/disable metrics and traces for each receiver independently, which is not possible without sublevels (without sublevels there is only one "Enabled" flag that applies both to metrics and traces of that receiver). However this exact same net result can be achieved by configuring the pipelines correctly, i.e. you can include that receiver in metrics pipeline but omit from traces pipeline and achieve the same effect.

The downside of using sublevels is that it results more verbose config file.